src/backends/lfc.rs

```diff
@@ -1,41 +1,106 @@
 
-use crate::args::BuildArgs;
-use crate::interface::Backend;
-use crate::lfc::LFCProperties;
+use serde_derive::Serialize;
+
+use crate::backends::{BatchBackend, BatchBuildResults, BuildCommandOptions, CommandSpec};
 use crate::package::App;
-use crate::util::command_line::run_and_capture;
+use crate::util::errors::BuildResult;
+
+pub struct LFC;
+
+impl LFC {
+    /// Do codegen for all apps in the batch result in parallel.
+    pub fn do_parallel_lfc_codegen(
+        options: &BuildCommandOptions,
+        results: &mut BatchBuildResults,
+        compile_target_code: bool,
+    ) {
+        results.par_map(|app| LFC::do_lfc_codegen(app, options, compile_target_code));
+    }
+
+    /// Do codegen for a single app.
+    fn do_lfc_codegen(
+        app: &App,
+        options: &BuildCommandOptions,
+        compile_target_code: bool,
+    ) -> BuildResult {
+        fs::create_dir_all(&app.output_root)?;
 
-pub struct LFC<'a> {
-    target: &'a App,
-    lfc: &'a LFCProperties,
+        let mut lfc_command = Command::new(&options.lfc_exec_path);
+        lfc_command.arg(format!(
+            "--json={}",
+            LfcJsonArgs::new(app, compile_target_code)
+        ));
+        crate::util::execute_command_to_build_result(lfc_command)
+    }
 }
 
-impl<'a> Backend<'a> for LFC<'a> {
-    fn from_target(target: &'a App, lfc: &'a LFCProperties) -> Self {
-        LFC { target, lfc }
+impl BatchBackend for LFC {
+    fn execute_command(&mut self, command: &CommandSpec, results: &mut BatchBuildResults) {
+        match command {
+            CommandSpec::Build(options) => {
+                LFC::do_parallel_lfc_codegen(options, results, options.compile_target_code)
+            }
+            CommandSpec::Update => todo!(),
+            CommandSpec::Clean => {
+                results.par_map(|app| {
+                    fs::remove_dir_all(app.src_gen_dir())?;
+                    Ok(())
+                });
+            }
+        }
     }
+}
 
-    fn build(&self, _config: &BuildArgs) -> io::Result<()> {
-        println!(
-            "--- Building main reactor: {}",
-            self.target.main_reactor.display()
-        );
-        // FIXME: What is this supposed to do? `lfc` does not have n `--output` argument
-        //  also. Why isnt the lfc from the CLI `-l` used.
-        let mut command = Command::new("lfc");
-        command.arg("--output-path");
-        command.arg("./");
-        command.arg(&self.target.main_reactor);
-        run_and_capture(&mut command).map(|_| ())
+/// Formats LFC arguments to JSON.
+#[derive(Serialize, Clone)]
+struct LfcJsonArgs<'a> {
+    /// Path to the LF source file containing the main reactor.
+    pub src: &'a Path,
+    /// Path to the directory into which build artifacts like
+    /// the src-gen and bin directory are generated.
+    pub out: &'a Path,
+    /// Other properties, mapped to CLI args by LFC.
+    pub properties: &'a HashMap<String, serde_json::Value>,
+    #[serde(skip)]
```

Comment:
I am not fully sure about this. We actually need to set the no-compile flag so flc currently doesn't invoke the toolchain.

Reply:
We need to set it but it isn't in the root of the JSON structure. See how `to_properties` works around this